File: app/src/main/java/com/jefferson/application/br/adapter/AppLockAdapter.java
package com.jefferson.application.br.adapter;

import android.app.Activity;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;
import com.jefferson.application.br.R;
import com.jefferson.application.br.database.AppLockDatabase;
import com.jefferson.application.br.model.AppModel;
import com.jefferson.application.br.service.AppLockService;
import com.jefferson.application.br.util.JDebug;
import com.jefferson.application.br.widget.LockCheck;
import java.util.ArrayList;
import java.util.HashMap;
import androidx.annotation.NonNull;

public class AppLockAdapter extends BaseAdapter {

    public SparseBooleanArray selectionArray = new SparseBooleanArray();
	private Activity activity;
	private LayoutInflater inflater = null;
    public ArrayList<AppModel> models;
    public AppLockDatabase database;
    public static AppLockService service;
    private boolean mutable;
    private HashMap<Integer, View> cachedViews;
    private int searchedItemPosition = -1;

	public AppLockAdapter(Activity mActivity, ArrayList<AppModel> models) {
        this.activity = mActivity;
        this.models = models;
        this.cachedViews = new HashMap<>();
		this.database = new AppLockDatabase(mActivity);
        syncSelection();

		inflater = (LayoutInflater) mActivity
			.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
	}

    public void animateSearchedItem(int x) {
        View view = cachedViews.get(x);

        if (view != null) {
            view.startAnimation(getBlinkAnimation());
        } else {
            JDebug.toast("View is NULL");
        }
    }

    private Animation getBlinkAnimation() {
        return AnimationUtils.loadAnimation(activity, R.anim.blink);         
    }

    public void setSearchedItem(int x) {
        this.searchedItemPosition = x;
    }

    public void clear() {
        this.models.clear();
        this.selectionArray.clear();
        this.cachedViews.clear();
        notifyDataSetInvalidated();
    }

    public void putDataSet(ArrayList<AppModel> newModels) {
        this.models = newModels;
        syncSelection();
        notifyDataSetChanged();
    }

    public void setMutable(boolean mutable) {
        this.mutable = mutable;
    }

    public boolean isMutable() {
        return mutable;
    }

	public View getView(final int position, View convertView, ViewGroup parent) {
        Holder holder = null;

        if (convertView == null) {
            convertView = inflater.inflate(R.layout.list_item, null);
            holder = new Holder();
            holder.imageView = (ImageView) convertView.findViewById(R.id.iconeApps);
            holder.textView = (TextView) convertView.findViewById(R.id.app_name);
            holder.lockCheck = (LockCheck) convertView.findViewById(R.id.check1);
            convertView.setTag(holder);
        } else {
            holder = (Holder) convertView.getTag();
        }
        AppModel info = models.get(position);
        @NonNull Drawable icon = info.icon;
		holder.imageView.setImageDrawable(icon);
		holder.textView.setText(info.appname);
		holder.lockCheck.setChecked(selectionArray.get(position, false));
        Animation animation = AnimationUtils.loadAnimation(activity, R.anim.zoom_in);
        animation.setAnimationListener(new Animation.AnimationListener() {

                @Override
                public void onAnimationStart(Animation p1) {

                }

                @Override
                public void onAnimationEnd(Animation anim) {
                    animateIfSearchedItem(position);
                }


                @Override
                public void onAnimationRepeat(Animation p1) {

                }
            }
        );
        convertView.startAnimation(animation);
        cachedViews.put(position, convertView);

        return convertView;
    }

    public int getItemHeight() {
        View item = cachedViews.get(0);
        return item == null ? 0 : item.getHeight();
    }

    static class Holder {
        ImageView imageView;
        TextView textView;
        LockCheck lockCheck;
    }
//    private Drawable getResizedDrawable(int position, View view) {
//
//        Drawable drawable = models.get(position).icon;
//        int previewWidth = 0;
//        int previewHeight = 0;
//        int mViewWidth = view.getWidth();
//        int mViewHeight = view.getHeight();
//        int newWidth = 0;
//        int newHeight = 0;
//
//        if (drawable != null && drawable.getIntrinsicWidth() > 0 && drawable.getIntrinsicHeight() > 0) { 
//            // the intrinsic dimensions can be -1 in some cases such as ColorDrawables which aim to fill 
//            // the whole View 
//            previewWidth = drawable.getIntrinsicWidth(); 
//            previewHeight = drawable.getIntrinsicHeight(); 
//        } 
//
//        final float widthScale = mViewWidth / (float) (previewWidth);
//        
//        if (widthScale != 1f) 
//            newWidth = Math.max((int)(widthScale * previewWidth), 1);
//        
//        final float heightScale = mViewHeight / (float) (previewHeight); 
//        
//        if (heightScale != 1f)
//            newHeight = Math.max((int)(heightScale * previewHeight), 1); 
//
//        // Define the Canvas and Bitmap the drawable will be drawn against 
//        final Canvas c = new Canvas(); 
//        Bitmap bitmap = ((BitmapDrawable)drawable).getBitmap(); 
//        c.setBitmap(bitmap); // Draw the scaled drawable into the final bitmap 
//        
//        if (drawable != null) {
//            drawable.setBounds(0, 0, newWidth, newHeight);
//            drawable.draw(c); 
//        }
//        return drawable;
//    }
//    
    private void animateIfSearchedItem(int position) {
        final View view = cachedViews.get(position);

        if (view == null) {
            return; //JDebug.toast("animateIfSearched: View is NULL");
        }

        if (searchedItemPosition == position) {
            view.post(new Runnable(){
                    @Override 
                    public void run() {
                        Animation blinkAnim = getBlinkAnimation();
                        //anim.setDuration(1000);
                        blinkAnim.setStartTime(1000);
                        view.startAnimation(blinkAnim);
                        searchedItemPosition = -1;
                    }
                }
            );
        }
    }

    private void syncSelection() {
        ArrayList<String> lockedPackages = database.getLockedPackages();

        for (int i = 0; i < getCount(); i++) {
            AppModel model = models.get(i);
            if (lockedPackages.contains(model.pname))
                selectionArray.put(i, true);
        }
    }

	public void toogleSelection(int position, View view) {
        setMutable(true);
		String pname = models.get(position).pname;
        boolean hasSelected = selectionArray.get(position, false);

        if (hasSelected) {
			selectionArray.delete(position);
			database.removeLockedApp(pname);
		} else {
			selectionArray.put(position, true);
			database.addLockedApp(pname);
		}
//        LockCheck lockView = view.findViewById(R.id.check1);
//        lockView.setChecked(!hasSelected);
	}

	public final int getCount() {
		return models.size();
	}

	public final Object getItem(int position) {
		return models.get(position);
	}

	public final long getItemId(int position) {
		return position;
	}

}

	

File: app/src/main/java/com/jefferson/application/br/adapter/AppLockAdapter.kt
/*
 * Copyright (C) 2023 Jefferson Calheiros


 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/
package com.jefferson.application.br.adapter

import android.content.Context
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import com.jefferson.application.br.R
import com.jefferson.application.br.database.AppLockDatabase
import com.jefferson.application.br.fragment.LockFragment
import com.jefferson.application.br.model.AppModel
import com.jefferson.application.br.util.JDebug
import com.jefferson.application.br.widget.LockCheck

@Suppress("NAME_SHADOWING")
class AppLockAdapter(private val fragment: LockFragment, var models: ArrayList<AppModel>) :
    BaseAdapter() {
    var scrollState: ScrollState = ScrollState.STOP
    private var selectionArray = SparseBooleanArray()
    var inflater: LayoutInflater
    var database: AppLockDatabase = AppLockDatabase(fragment.context)
    var isMutable = false
    private val cachedViews: HashMap<Int, View?> = HashMap()
    private var searchedItemPosition = -1

    init {
        syncSelection()
        inflater = fragment.context
            ?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    }

    fun animateSearchedItem(x: Int) {
        val view = cachedViews[x]
        if (view != null) {
            view.startAnimation(blinkAnimation)
        } else {
            JDebug.toast("View is NULL")
        }
    }

    private val blinkAnimation: Animation
        get() = AnimationUtils.loadAnimation(fragment.context, R.anim.blink)

    fun setSearchedItem(x: Int) {
        searchedItemPosition = x
    }

    fun clear() {
        models.clear()
        selectionArray.clear()
        cachedViews.clear()
        notifyDataSetInvalidated()
    }

    fun putDataSet(newModels: ArrayList<AppModel>) {
        models = newModels
        syncSelection()
        notifyDataSetChanged()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
        var convertView = convertView
        val holder: Holder?

        if (convertView == null) {
            convertView = inflater.inflate(R.layout.list_item, parent, false)
            holder = Holder()
            holder.imageView = convertView.findViewById<View>(R.id.iconeApps) as ImageView
            holder.textView = convertView.findViewById<View>(R.id.app_name) as TextView
            holder.lockCheck = convertView.findViewById<View>(R.id.check1) as LockCheck
            convertView.tag = holder
        } else {
            holder = convertView.tag as Holder
        }
        val info = models[position]
        val icon = info.icon
        holder.imageView?.setImageDrawable(icon)
        holder.textView?.text = info.packageName
        holder.lockCheck?.isChecked = selectionArray[position, false]
        if (scrollState != ScrollState.STOP) {
            val animation = AnimationUtils.loadAnimation(
                fragment.context, when (scrollState) {
                    ScrollState.UP -> R.anim.app_slide_bottom
                    ScrollState.DOWN -> R.anim.app_slide_up
                    else -> 0
                }
            )
            animation.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(p1: Animation) {}
                override fun onAnimationEnd(anim: Animation) {
                    animateIfSearchedItem(position)
                }
                override fun onAnimationRepeat(p1: Animation) {}
            }
            )

            val middleOfList = (fragment.lastVisibleItem - fragment.firstVisibleItem) / 2
            //FIX: wrong animation on top and bottom
            if (scrollState == Companion.ScrollState.UP &&
                fragment.firstVisibleItem > position - middleOfList
                || scrollState == Companion.ScrollState.DOWN &&
                (fragment.firstVisibleItem < position - middleOfList)) {

                convertView?.startAnimation(animation)
            }
        }

        cachedViews[position] = convertView
        return convertView
    }

    val itemHeight: Int
        get() {
            val item = cachedViews[0]
            return item?.height ?: 0
        }

    internal class Holder {
        var imageView: ImageView? = null
        var textView: TextView? = null
        var lockCheck: LockCheck? = null
    }

    private fun animateIfSearchedItem(position: Int) {
        val view = cachedViews[position]
            ?: return  //JDebug.toast("animateIfSearched: View is NULL");
        if (searchedItemPosition == position) {
            view.post {
                val blinkAnim: Animation = blinkAnimation
                //anim.setDuration(1000);
                blinkAnim.startTime = 1000
                view.startAnimation(blinkAnim)
                searchedItemPosition = -1
            }
        }
    }

    private fun syncSelection() {
        val lockedPackages = database.lockedPackages
        for (i in 0 until count) {
            val model = models[i]
            if (lockedPackages.contains(model.name)) selectionArray.put(i, true)
        }
    }

    fun toggleSelection(position: Int) {
        isMutable = true
        val name = models[position].name
        val hasSelected = selectionArray[position, false]
        if (hasSelected) {
            selectionArray.delete(position)
            database.removeLockedApp(name)
        } else {
            selectionArray.put(position, true)
            database.addLockedApp(name)
        }
        //        LockCheck lockView = view.findViewById(R.id.check1);
//        lockView.setChecked(!hasSelected);
    }

    override fun getCount(): Int {
        return models.size
    }

    override fun getItem(position: Int): Any {
        return models[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    companion object {
        enum class ScrollState {
            UP,
            DOWN,
            STOP,
        }
    }
}